703 field permission guide

bradfordcondon opened the pull request:
<!--- Thank you for contributing! -->
<!--- Provide a general summary of your changes in the Title above -->
<!--- See our Contribution Guidelines here:
          https://github.com/tripal/tripal/blob/7.x-3.x/CONTRIBUTING.md -->


<!---  Please set the header below based on the PR type:
# New Feature
# Bux Fix
# Documentation  --->

# Documentation

Issue #703

## Description

Provides documentation for setting individual field permissions using the field permissions module.

## Testing

Ideally you will build the documentation locally (`make html` in the docs folder, open the build in `_build`) and then follow the guide to configure a field permission.  If you havent built our docs before you'll need to [install sphinx first](http://www.sphinx-doc.org/en/master/usage/installation.html)


## images

![screen shot 2018-11-09 at 4 10 51 pm](https://user-images.githubusercontent.com/7063154/48288428-14066680-e43a-11e8-9ed6-73ec247c8d72.png)

bradfordcondon committed:
add field permission page

bradfordcondon committed:
fix file path and header length

bradfordcondon committed:
one more sentence

bradfordcondon committed:
new image

bradfordcondon commented:
question to ask yourself if you review: is this document in the right place :P


edit: now that im thinking about this, it belongs in the user guide instead, iun the "set up example genomic site" for sure.

almasaeed2010 commented on docs/dev_guide/field_permissions.rst:
```diff
@@ -0,0 +1,50 @@
+
+.. image:: ./field_permissions.1.cross_ref_GA.png
+
+First, navigate to the bundle field configuration page via **Admin --> Structure --> Tripal Content --> Genome Assembly**.  For each field we want to hide, we must configure the field instance settings individually.  Click **Edit** for the Cross Reference field, and scroll down to **CROSS REFERENCE FIELD SETTINGS**.
```
Admin --> Structure --> Tripal Content --> Genome Assembly --> Manage Fields

Missing `Manage Fields`

almasaeed2010 reviewed (approved):
Followed this guide on our live site to customize field permissions. Clear and very helpful! Thanks @bradfordcondon

bradfordcondon commented:
thanks for reviewing.  Doyou agree this belongs in user guide instead of dev guide?

almasaeed2010 commented:
I do. There is no programming at all in this. It's part of using Tripal rather than developing for Tripal.

bradfordcondon commented:
agreed. ill commit to move it.

bradfordcondon committed:
move to user guide

bradfordcondon commented:
ok i stuck it in the user guide instead.  i still am not sure if i like where it is, but its an improvement.

spficklin commented:
Some suggestions for this document.
1)  Can we move it under the header 'Working with Content Types"
2)  I think since this is meant for end-users (not developers) we need to adjust some vocabulary.
    a)  The word bundle should be renamed 'content types'.  I don't think we use the word bundle for users..
    b)  It refers to the HQ module, but that's not been introduced in our tutorial prior to this document.  Perhaps we need an introduction here to the HQ module and how this field_permission module integrates with it? Or we can just leave out the HQ sentences?
    c)  So this sentence bothers me:  "Some fields are just not intuitive without some Chado experience: for example, the Cross-Reference field."    The point of fields was to take Chado out of the picture. A cross reference should be understandable without knowing anything about Chado... right?

So, one last thing to consider.... The web services do not support field_permissions I don't think.  So, if a field is hidden on a page it should also be hidden in web services, right?  I know this is just documentation and not code, but we should probably test it to see if web services is also affected.

bradfordcondon commented:
Thanks a lot for your feedback.  I wrote it on my blog thinking of developers but you are right i should re-think it with end users more in mind now that i've moved it there.


I moved it where you suggested, renamed bundles to content types.


>Or we can just leave out the HQ sentences?

I tried removing HQ all-together.  What do you think?  Should we add a `..note` that says if you are interested in having users submitting content, check out Tripal HQ?
 
>A cross reference should be understandable without knowing anything about Chado... right?

The field might be understandable, but not the **Widget** i would say?  Really what i mean is you need to add stuff to another table first... How about: "Some fields require prior insertion of data elsewhere: for example, the Cross-Reference field."

(side note, i just tried creating an analysis with a cross ref to the db AGL of the nonexistent accession "waffle".  the content got created, but without a dbxref.  No warning, no validation failure. This is what im getting at, a user might get confounded that they need to insert the accession first, and maybe they dont even have permission to do so.  Thats why we want to hide this widget from them.)


>The web services do not support field_permissions I don't think. So, if a field is hidden on a page it should also be hidden in web services, right? I know this is just documentation and not code, but we should probably test it to see if web services is also affected.

This is interesting.  The desired behavior would be to display the field if its set to display to an anonymous user (logged in user?)?  And the other purpose of field permissions, setting it for the widget only, doesnt effect web services...

bradfordcondon commented:
Looks like setting field permissions has no effect on web services: the field shows up  no matter what.

The main purpose of field permissions is (in my case) to hide confusing widgets from curators submitting content. Hiding content for non-admins was a bonus.  I would say we could keep as is and add a warning that content wont be hidden from web-services in the guide, but then we still have that warning to deal with....

Here's the warning i get:

```php
Notice: Undefined offset: 0 in TripalContentService_v0_1->addEntityField() (line 383 of /Users/bc/tripal/sites/all/modules/tripal/tripal_ws/includes/TripalWebService/TripalContentService_v0_1.inc).
Notice: Undefined offset: 0 in TripalContentService_v0_1->addEntityField() (line 396 of /Users/bc/tripal/sites/all/modules/tripal/tripal_ws/includes/TripalWebService/TripalContentService_v0_1.inc).

```

bradfordcondon committed:
revised basedon spficklin suggestions

bradfordcondon commented:
easy to resolve the warning by adding this to the web services class just prior to here: 

https://github.com/tripal/tripal/blob/bfca9765a5ef7aece7b4355c886d91513a89cadb/tripal_ws/includes/TripalWebService/TripalContentService_v0_1.inc#L383


```php
  if (empty($values)){
        return;
      }

      // If the value is an array and this is the field page then all of those
      // key/value pairs should be added directly to the response.
      if (is_array($values[0])) {
```

adding this also causes the ws to work how i'd expect: it isnt returned since we didnt have the permission enabled for an anonymous user.

probably also need to add a case for where cardinality != 1....  what do you think?  This would make the ws behave how i expect (field isnt shown on web services), and it resolves the php warning.

spficklin commented:
I like the changes!  I think we can merge, but I think we should add a new issue that requests support for the field_permissions within the web services.  If we recommend the module we should be sure to support it.

bradfordcondon commented:
awesome, thanks!  Ill make a separate issue for supporting it in web services so we can take more care in applying the fix, as you suggest.

bradfordcondon commented:
im going to add in a note referring users to tripal HQ for more information about users submitting content, then merge.  thanks for hte reviews both of you.

bradfordcondon committed:
add tripal HQ box
